### PR TITLE
Fix autoplay

### PR DESCRIPTION
**Pull request by nickcuper**
This change is [Reviewable](https://reviewable.io/reviews/openplayerjs/openplayerjs/110)

### CODE REVIEW

File: src/js/utils/media.ts [hunk 1]
```diff
@@ -96,50 +95,32 @@ export function isAutoplaySupported(media: HTMLMediaElement, autoplay: (n: any)
...
-                try {
-                    autoplay(false);
-                    muted(false);
-                    callback();
-                } catch (e) {
-                    throw new Error('Cannot autoplay');
-                }
+                autoplay(false);
+                muted(false);
+                callback();
```
Comment (rafa8626) [inline src/js/utils/media.ts]: Shouldn’t we be returning the callback result here as well?

File: src/js/utils/media.ts [hunk 2]
```diff
@@ -96,50 +95,32 @@ export function isAutoplaySupported(media: HTMLMediaElement, autoplay: (n: any)
...
-            media.pause();
-            muted(false);
-            callback();
-        } catch (e) {
-            throw new Error('Attempt to autoplay failed');
-        }
+        autoplay(!media.paused || 'Promise' in window && playPromise instanceof Promise);
+        media.pause();
+        muted(false);
+        callback();
```
Comment (rafa8626) [inline src/js/utils/media.ts]: And here as well?

File: src/js/player.ts [hunk 3]
```diff
@@ -344,15 +344,15 @@ class Player {
...
      * @memberof Player
      */
-    public play(): void {
+    public play(): Promise<void> | void {
         if (this.media && !this.media.loaded) {
             this.media.load();
             this.media.loaded = true;
         }
         if (this.adsInstance) {
             this.adsInstance.play();
```
Comment (rafa8626) [inline src/js/player.ts]: This needs the same workflow than the regular media to be consistent

File: src/js/utils/media.ts (hunk 2 above)
Comment (nickcuper) [reply src/js/utils/media.ts]: not needed here

File: src/js/utils/media.ts (hunk 1 above)
Comment (nickcuper) [reply src/js/utils/media.ts]: not necessary here

**Comment by rafa8626** [comment]
@nickcuper one of the tests is failing; just check why this is happening and I'll merge your changes

File: src/js/player.ts (hunk 3 above)
Comment (nickcuper) [reply src/js/player.ts]: done

**Comment by nickcuper** [comment]
> Last question: since you integrated the Promise into native, it doesn't affect the other media types, such as hls.js or dash.js, correct?

i have checked these types
```html
<option value="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/TearsOfSteel.mp4" selected>MP4</option>
                <option value="https://upload.wikimedia.org/wikipedia/commons/2/22/Volcano_Lava_Sample.webm">WebM</option>
                <option value="https://player.webvideocore.net/CL1olYogIrDWvwqiIKK7eLBkzvO18gwo9ERMzsyXzwt_t-ya8ygf2kQBZww38JJT/8i4vvznv8408.m3u8">HLS</option>
                <option value="https://dash.akamaized.net/envivio/EnvivioDash3/manifest.mpd">M(PEG)-DASH</option>
```

it doesn't

But i found others issue for some edge cases (i suggest to fix them in separate tasks, cuz it doesn't related to autoplay in general)
1. Enabled AdBlock in chrome throws unhandled error
2. Safari - dash.js doesn't works at all, looks like it problem on dash.js level, i have checked their demo and got same result
3. Safari does not supporting webm type (https://caniuse.com/#feat=webm)

**Comment by rafa8626** [comment]
@nickcuper Thanks for the update, and I agree; just FYI dash and webm will never work. The one with AdBlocker we can treat it definitely as a separate issue. 

So once the tests pass I can proceed to merge your branch

**Comment by nickcuper** [comment]
fix the test, before it was incorrect

**Comment by rafa8626** [comment]
@nickcuper thank you so much for your work on this!

**Comment by nickcuper** [comment]
> @nickcuper thank you so much for your work on this!

no problem

**Comment by rafa8626** [comment]
And just FYI you’re honorably the first contributor to the project so congrats!

**Pull request merged**